akka-stream/src/main/scala/akka/stream/impl/HotPatch.scala

```diff
@@ -0,0 +1,68 @@
+package akka.stream.impl
+
+import akka.event.Logging
+import akka.stream.Attributes.Attribute
+import akka.stream._
+import akka.stream.impl.HotPatch.HotPatchAttribute
+import akka.stream.stage.{ GraphStageLogic, GraphStage }
+
+/**
+ * INTERNAL API
+ */
+private[akka] object HotPatch {
+
+  case class HotPatchAttribute[I, O](placeholder: HotPatch[I, O], replacement: GraphStage[FlowShape[I, O]])
+    extends Attribute
+
+  def patch[I, O](placeholder: HotPatch[I, O], replacement: GraphStage[FlowShape[I, O]]): Attributes = {
+    Attributes(HotPatchAttribute(placeholder, replacement))
+  }
+
+}
+
+/**
+ * INTERNAL API
+ *
+ * The purpose of this stage is to allow prefusing a large graph by placing this placeholder at a position where
+ * a dynamically injected Flow is required. The actual Flow can be injected by an Attribute just before materialization.
+ */
+private[akka] class HotPatch[I, O] extends GraphStage[FlowShape[I, O]] {
```

Comment:
I guess this would work for arbitrary shapes (with some magic to produce the necessary Shape instance—or by taking it as an argument).